### PR TITLE
simple optimization to ForceField class

### PR DESCRIPTION
**Pull request by tristanic**
For force field libraries containing large numbers of residue templates (about 15,000 in my current case), `ForceField.__init__()` spends some 20-30% of its time (4 of a total 20 seconds) just on `ForceField._TemplateData.getAtomIndexByName()`. A few simple adjustments cut that by a factor of 10.

### CODE REVIEW

**Comment by peastman** [comment]
Thanks!  It looks like some tests are failing.

**Comment by tristanic** [comment]
Hmm - will double-check in the morning. I don’t routinely build OpenMM, so just hand-copied my edits over from my working 7.3.0 environment. Probably made a typo somewhere.

 
 
Tristan Croll
Research Fellow
Cambridge Institute for Medical Research
University of Cambridge CB2 0XY

**Comment by tristanic** [comment]
Aargh! Sorry - the `addAtom()` method was a new addition I made while copying over from my working code, and of course I managed to add an out-by-one error. I've now rebuilt locally, and checked that all relevant Python tests are passing.

**Comment by tristanic** [comment]
Sorry for being a complete n00b with GitHub, but that last commit was supposed to be on a different branch for a separate pull request. Noticed a few `ResourceWarning`s while running the Python tests - the ones in the test code are fairly harmless, but `PDBxFile`  and `GromacsGroFile` were both leaving their input files open indefinitely.

File: wrappers/python/simtk/openmm/app/forcefield.py [hunk 1]
```diff
@@ -263,17 +263,17 @@ def loadFile(self, files):
...
-                    for atom in residue.findall('Atom'):
+                    atomIndices = template.atomIndices
+                    for ia, atom in enumerate(residue.findall('Atom')):
                         params = {}
-                        for key in atom.attrib:
-                            if key not in ('name', 'type'):
-                                params[key] = _convertParameterToNumber(atom.attrib[key])
-                        atomName = atom.attrib['name']
+                        a_attrib = atom.attrib
+                        atomName = a_attrib.pop('name')
```
Comment (peastman) [inline wrappers/python/simtk/openmm/app/forcefield.py]: This line makes me a little nervous.  We load the XML into the etree data structure, then copy information from it into our own data structures.  But here you're directly modifying the etree data structures.  That's the sort of thing that could easily lead to bugs later on, for example if someone writes more code that looks at those structures and expects them to still reflect the content of the file.

What about keeping the parsing code as it was before and just adding the single line `template.atomIndices = atomIndices` at the end?

**Comment by tristanic** [comment]
That’s fine by me. In my trials it shaved a couple percent off the load time, but that’s pretty minor in the grand scheme of things. Won’t be able to make the edit until tomorrow morning (it’s 6:45pm here) - if you want to do it, feel free.

**Comment by peastman** [comment]
Look good.

**Pull request merged**